### ISSUE TITLE
Moderated Sessions

### ISSUE THREAD

**Issue by vitorenesduarte** [gravitational/predicate-lang#30]
I opened this issue to try to discuss how to best proceed with #26.

(This is issue is mostly self contained, even though it makes a reference to #29.)

## Teleport RBAC today

For a user to access a node, it must have a role with one the following (in the `allow` section):
- `node_labels`
- `node_labels` and `require_session_join`
- `join_sessions`

### `require_session_join` is not an action

`require_session_join` doesn't grant any node access, it simply places restrictions on it.
This means that `require_session_join` is not action (see #29 for a definition of action).

For this reason, it doesn't make sense to allow/deny it.
This explains why the following has no meaning in Teleport RBAC today:

```yaml
kind: role
metadata:
  name: prod-access
spec:
  deny:
    require_session_join:
        ...
```

### What this means for #26

In #26 we're trying to introduce both `StartSession` and `JoinSession` in order to mimic what we have today as `require_join_session` and `join_sessions` (respectively).

Since `require_join_session` does not represent an action, I think that we should remove `StartSession` from #26 and only add `JoinSession` for now.

### So how do we restrict node access like we do with `require_join_session`?

IDK what's the best way yet but maybe `require_join_session` should be specified using session options.

Session options seem to be properties that must be maintained throughout the session, not when the initial action ("node access", in this case) is performed.

This seems to be true for at least `max_session_ttl` and `recording_mode`.

I leave below another two arguments that seem to hint that `require_join_session` restrictions should indeed be session options.

#### 1. `require_join_session` across roles are combined like options

The [docs](https://goteleport.com/docs/access-controls/guides/moderated-sessions/#combining-policies) say the following about __Combining Policies__:

> The authorizer applies require policies within a role together with an OR operator and the policies from each role with an AND operator. In practice, this means that for every role with at least one require policy, one of its policies must be met before a session can be started.

This is exactly what we do with options: if the same option is defined in multiple polices, we apply the logical `AND`, and that seems to be what's done with `require_join_session`.

This is not the case for allow rules where we apply the logical `OR`.
So again, it seems that the allow rules is not the best place for `require_join_session`.

#### 2. `require_join_session.on_leave` combines like `options.recording_mode`

The [docs](https://goteleport.com/docs/access-controls/guides/moderated-sessions/#on-leave) say the following about __On Leave__:

> By default, Teleport treats an empty string in this field as the same as terminate. That is, the session is terminated instantly and all participants are disconnected.
> If all require policies attached to the session owner are set to "pause", the session will instead pause but the session will remain open. 

IIUC if one role says `on_leave: "pause"` and another says `on_leave: "terminate"`, they are combined as `on_leave: "terminate"`.

This feels like an option where:
- `(0, "pause")`,
- `(1, "terminate")`,
- and we take the highest value,

which is what we're [doing today in predicate with `recording_mode`](https://github.com/gravitational/predicate-lang/blob/2356b3a5e1536c1ad9fbf935197389754255e55c/predicate/solver/teleport.py#L41-L43) where:
- `(0, "best_effort")`,
- `(1, "strict")`,
- and we also take the highest value.

**Comment by xacrimon** [comment]
I agree with this analysis and that we can leave out startsession for now. We can add it going forward. What do you think about this options idea? @klizhentas
